<a href="https://colab.research.google.com/github/MilanL96/Intro-to-Data-Science-Assignment/blob/main/Component_1_Milan_Logman_17002423.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **England School Information**

In [134]:
import pandas as pd

In [135]:
pd.read_csv('/content/drive/MyDrive/OFSTED Data for Assignment/england_school_information.csv')

URN            LANAME   LA  ...  GENDER            RELCHAR          ADMPOL
0      100000    City of London  201  ...   Mixed  Church of England  Not applicable
1      100001    City of London  201  ...   Girls               None  Not applicable
2      100002    City of London  201  ...   Mixed  Church of England  Not applicable
3      100003    City of London  201  ...    Boys               None  Not applicable
4      100008            Camden  202  ...   Mixed     Does not apply  Not applicable
...       ...               ...  ...  ...     ...                ...             ...
25287  148179       East Sussex  845  ...   Mixed     Does not apply   Non-selective
25288  148185      Lincolnshire  925  ...   Mixed               None   Non-selective
25289  148186  Stockton-on-Tees  808  ...   Mixed               None             NaN
25290  148187        Birmingham  330  ...   Mixed               None             NaN
25291  148193     Milton Keynes  826  ...   Mixed     Does not apply  Not applicable

[25292 rows x 24 columns]

In [136]:
esi = pd.read_csv('/content/drive/MyDrive/OFSTED Data for Assignment/england_school_information.csv')

In [137]:
esi.head()

URN          LANAME   LA  ...  GENDER            RELCHAR          ADMPOL
0  100000  City of London  201  ...   Mixed  Church of England  Not applicable
1  100001  City of London  201  ...   Girls               None  Not applicable
2  100002  City of London  201  ...   Mixed  Church of England  Not applicable
3  100003  City of London  201  ...    Boys               None  Not applicable
4  100008          Camden  202  ...   Mixed     Does not apply  Not applicable

[5 rows x 24 columns]

In [138]:
esi = esi[(esi.ISSECONDARY>0)]

In [139]:
esi = esi.drop(columns=['LANAME', 'ESTAB', 'LAESTAB', 'SCHNAME', 'STREET', 'LOCALITY', 'ADDRESS3',	'TOWN', 'POSTCODE', 'SCHSTATUS', 'OPENDATE', 'CLOSEDATE', 'MINORGROUP', 'SCHOOLTYPE', 'ISPOST16', 'ISPRIMARY', 'ISSECONDARY','AGELOW','AGEHIGH','ADMPOL'])

In [140]:
esi

URN   LA GENDER            RELCHAR
1      100001  201  Girls               None
2      100002  201  Mixed  Church of England
3      100003  201   Boys               None
41     100049  202  Mixed     Does not apply
42     100050  202  Girls     Does not apply
...       ...  ...    ...                ...
25276  148135  391  Mixed     Does not apply
25287  148179  845  Mixed     Does not apply
25288  148185  925  Mixed               None
25289  148186  808  Mixed               None
25290  148187  330  Mixed               None

[6396 rows x 4 columns]

# **England Census**

In [141]:
ec = pd.read_excel('/content/drive/MyDrive/OFSTED Data for Assignment/england_census.xlsx')

In [142]:
ec.head()

URN     LA ESTAB  ... NUMFSM  NUMFSMEVER  PNUMFSMEVER
0  100000  201.0  3614  ...   32.0        38.0         16.2
1  100001  201.0  6005  ...    NaN         NaN          NaN
2  100002  201.0  6006  ...    NaN         NaN          NaN
3  100003  201.0  6007  ...    NaN         NaN          NaN
4  100005  202.0  1048  ...   27.0         NaN          NaN

[5 rows x 22 columns]

In [143]:
ec['PNUMFSM'] = (ec.NUMFSM/ec.NOR)*100

In [144]:
ec.head()

URN     LA ESTAB  ... NUMFSMEVER  PNUMFSMEVER    PNUMFSM
0  100000  201.0  3614  ...       38.0         16.2  11.228070
1  100001  201.0  6005  ...        NaN          NaN        NaN
2  100002  201.0  6006  ...        NaN          NaN        NaN
3  100003  201.0  6007  ...        NaN          NaN        NaN
4  100005  202.0  1048  ...        NaN          NaN  20.300752

[5 rows x 23 columns]

In [145]:
ec=ec.drop(columns=['ESTAB','SCHOOLTYPE','NOR','NORG','NORB','TSENELSE','TSENELK','NUMEAL','NUMENGFL','NUMUNCFL','NUMFSM','NUMFSMEVER'])

In [146]:
ec['PNUMFSM']=ec['PNUMFSM'].round(decimals=1)

In [147]:
ec

URN     LA  PNORG  PNORB  ...  PNUMENGFL  PNUMUNCFL  PNUMFSMEVER  PNUMFSM
0      100000  201.0   49.5   50.5  ...       50.2        0.4         16.2     11.2
1      100001  201.0  100.0    0.0  ...        NaN        NaN          NaN      NaN
2      100002  201.0   41.3   58.7  ...        NaN        NaN          NaN      NaN
3      100003  201.0    0.0  100.0  ...        NaN        NaN          NaN      NaN
4      100005  202.0   47.4   52.6  ...       30.8        0.0          NaN     20.3
...       ...    ...    ...    ...  ...        ...        ...          ...      ...
23937  146826  845.0   45.4   54.6  ...       98.5        0.0         36.1     22.8
23938  146827  928.0   50.2   49.8  ...       61.9        0.0         29.0     18.7
23939     NAT    NaN   49.0   51.0  ...       78.6        0.2         23.0     15.8
23940     NAT    NaN   49.8   50.2  ...       82.7        0.5         27.7     14.1
23941     NAT    NaN   27.8   72.2  ...       85.0        0.3         50.3     38.6

[23942 rows x 11 columns]

# **England abs**

In [148]:
abs = pd.read_excel('/content/drive/MyDrive/OFSTED Data for Assignment/england_abs.xlsx')

In [149]:
abs.head()

LA ESTAB       URN  PERCTOT PPERSABS10
0  201  3614  100000.0      2.6        3.4
1  202  2000  136807.0      4.3         11
2  202  2001  139837.0      3.8        4.4
3  202  2002  140686.0      4.6       12.1
4  202  2019  100008.0      4.2        7.3

In [150]:
abs=abs.drop(columns=['ESTAB'])

In [151]:
abs

LA       URN  PERCTOT PPERSABS10
0      201  100000.0      2.6        3.4
1      202  136807.0      4.3         11
2      202  139837.0      3.8        4.4
3      202  140686.0      4.6       12.1
4      202  100008.0      4.2        7.3
...    ...       ...      ...        ...
21212  938  126170.0     17.9       41.3
21213  938  135814.0     14.1       57.8
21214  NAT       NaN      4.0        8.2
21215  NAT       NaN      5.5       13.7
21216  NAT       NaN     10.1       28.8

[21217 rows x 4 columns]

# **KS4 Provisional**

In [152]:
ks4prov = pd.read_excel('/content/drive/MyDrive/OFSTED Data for Assignment/england_ks4provisional.xlsx')

In [153]:
ks4prov.head()

RECTYPE    LEA   ESTAB       URN  ... PTEBACC_94  PTEBACC_95 PT5EM_94  P8_BANDING
0        1  201.0  6007.0  100003.0  ...        NaN         NaN      NaN         NaN
1        1  201.0  6005.0  100001.0  ...        NaN         NaN      NaN         NaN
2        4  201.0     NaN       NaN  ...        NaN         NaN      NaN         NaN
3        1  202.0  4285.0  100053.0  ...      0.368       0.265    0.625           3
4        2  202.0  7137.0  100092.0  ...          0           0        0           5

[5 rows x 25 columns]

In [154]:
ks4prov=ks4prov.drop(columns=['RECTYPE','ESTAB','SCHNAME','ICLOSE','NFTYPE','P8MEA_ORIG','P8CILOW_ORIG','P8CIUPP_ORIG','TP8ADJ'])

In [155]:
ks4prov

LEA       URN      TPUP  ... PTEBACC_95 PT5EM_94 P8_BANDING
0     201.0  100003.0       NaN  ...        NaN      NaN        NaN
1     201.0  100001.0       NaN  ...        NaN      NaN        NaN
2     201.0       NaN       NaN  ...        NaN      NaN        NaN
3     202.0  100053.0     136.0  ...      0.265    0.625          3
4     202.0  100092.0      17.0  ...          0        0          5
...     ...       ...       ...  ...        ...      ...        ...
5647  938.0  126137.0       NaN  ...        NaN      NaN        NaN
5648  938.0  139109.0     184.0  ...      0.255    0.712          3
5649  938.0       NaN    8002.0  ...      0.162    0.614          2
5650    NaN       NaN  605874.0  ...      0.157    0.554        NaN
5651    NaN       NaN  542831.0  ...       0.17      0.6        NaN

[5652 rows x 16 columns]

# **KS4 Pupdest**

In [156]:
ks4pup = pd.read_excel('/content/drive/MyDrive/OFSTED Data for Assignment/england_ks4-pupdest.xlsx')

In [157]:
ks4pup.head()

RECTYPE    LEA  ...  COHORT_NONDIS_17  OVERALL_DESTPER_NONDIS_17
0       1  202.0  ...               NaN                        NaN
1       1  202.0  ...                61                       0.92
2       1  202.0  ...                86                       0.97
3       1  202.0  ...                36                       0.97
4       1  202.0  ...                83                       0.98

[5 rows x 82 columns]

In [158]:
ks4pup=ks4pup[['URN','LEA','OVERALL_DESTPER','NOT_SUSTAINEDPER','UNKNOWNPER','OVERALL_DESTPER_DIS','NOT_SUSTAINEDPER_DIS','UNKNOWNPER_DIS']]

In [159]:
ks4pup

URN    LEA  ... NOT_SUSTAINEDPER_DIS UNKNOWNPER_DIS
0     137181.0  202.0  ...                 0.06           0.01
1     100049.0  202.0  ...                 0.11           0.01
2     100050.0  202.0  ...                 0.04           0.01
3     100051.0  202.0  ...                 0.09           0.06
4     100052.0  202.0  ...                 0.13              0
...        ...    ...  ...                  ...            ...
3872  136114.0  938.0  ...                 0.07              0
3873  126170.0  938.0  ...                 SUPP           SUPP
3874  135814.0  938.0  ...                 SUPP           SUPP
3875       NaN  938.0  ...                  0.1           0.02
3876       NaN    NaN  ...                  0.1           0.01

[3877 rows x 8 columns]

# **Merge Dataframes**

In [160]:
df=pd.DataFrame.merge(esi,ec,on='URN',how='left')

In [161]:
df

URN  LA_x GENDER  ... PNUMUNCFL  PNUMFSMEVER  PNUMFSM
0     100001   201  Girls  ...       NaN          NaN      NaN
1     100002   201  Mixed  ...       NaN          NaN      NaN
2     100003   201   Boys  ...       NaN          NaN      NaN
3     100049   202  Mixed  ...       0.1         64.5     41.2
4     100050   202  Girls  ...       0.9         49.6     26.9
...      ...   ...    ...  ...       ...          ...      ...
6391  148135   391  Mixed  ...       NaN          NaN      NaN
6392  148179   845  Mixed  ...       NaN          NaN      NaN
6393  148185   925  Mixed  ...       NaN          NaN      NaN
6394  148186   808  Mixed  ...       NaN          NaN      NaN
6395  148187   330  Mixed  ...       NaN          NaN      NaN

[6396 rows x 14 columns]

In [162]:
df=pd.DataFrame.merge(df,abs,on='URN',how='left')

In [163]:
df

URN  LA_x GENDER            RELCHAR  ...  PNUMFSM   LA  PERCTOT  PPERSABS10
0     100001   201  Girls               None  ...      NaN  NaN      NaN         NaN
1     100002   201  Mixed  Church of England  ...      NaN  NaN      NaN         NaN
2     100003   201   Boys               None  ...      NaN  NaN      NaN         NaN
3     100049   202  Mixed     Does not apply  ...     41.2  202      6.9        18.8
4     100050   202  Girls     Does not apply  ...     26.9  202      5.3        12.5
...      ...   ...    ...                ...  ...      ...  ...      ...         ...
6391  148135   391  Mixed     Does not apply  ...      NaN  NaN      NaN         NaN
6392  148179   845  Mixed     Does not apply  ...      NaN  NaN      NaN         NaN
6393  148185   925  Mixed               None  ...      NaN  NaN      NaN         NaN
6394  148186   808  Mixed               None  ...      NaN  NaN      NaN         NaN
6395  148187   330  Mixed               None  ...      NaN  NaN      NaN         NaN

[6396 rows x 17 columns]

In [164]:
df=pd.DataFrame.merge(df,ks4prov,on='URN',how='left')

In [165]:
df

URN  LA_x GENDER  ... PTEBACC_95  PT5EM_94  P8_BANDING
0     100001   201  Girls  ...        NaN       NaN         NaN
1     100002   201  Mixed  ...        NaN       NaN         NaN
2     100003   201   Boys  ...        NaN       NaN         NaN
3     100049   202  Mixed  ...      0.141     0.533           3
4     100050   202  Girls  ...      0.326     0.749           1
...      ...   ...    ...  ...        ...       ...         ...
6391  148135   391  Mixed  ...        NaN       NaN         NaN
6392  148179   845  Mixed  ...        NaN       NaN         NaN
6393  148185   925  Mixed  ...        NaN       NaN         NaN
6394  148186   808  Mixed  ...        NaN       NaN         NaN
6395  148187   330  Mixed  ...        NaN       NaN         NaN

[6396 rows x 32 columns]

In [166]:
df=pd.DataFrame.merge(df,ks4pup,on='URN',how='left')

In [167]:
df

URN  LA_x  ... NOT_SUSTAINEDPER_DIS UNKNOWNPER_DIS
0     100001   201  ...                  NaN            NaN
1     100002   201  ...                  NaN            NaN
2     100003   201  ...                  NaN            NaN
3     100049   202  ...                 0.11           0.01
4     100050   202  ...                 0.04           0.01
...      ...   ...  ...                  ...            ...
6391  148135   391  ...                  NaN            NaN
6392  148179   845  ...                  NaN            NaN
6393  148185   925  ...                  NaN            NaN
6394  148186   808  ...                  NaN            NaN
6395  148187   330  ...                  NaN            NaN

[6396 rows x 39 columns]

In [168]:
df

URN  LA_x  ... NOT_SUSTAINEDPER_DIS UNKNOWNPER_DIS
0     100001   201  ...                  NaN            NaN
1     100002   201  ...                  NaN            NaN
2     100003   201  ...                  NaN            NaN
3     100049   202  ...                 0.11           0.01
4     100050   202  ...                 0.04           0.01
...      ...   ...  ...                  ...            ...
6391  148135   391  ...                  NaN            NaN
6392  148179   845  ...                  NaN            NaN
6393  148185   925  ...                  NaN            NaN
6394  148186   808  ...                  NaN            NaN
6395  148187   330  ...                  NaN            NaN

[6396 rows x 39 columns]

In [169]:
df=df.drop(columns=['LEA_x','LA','LA_y','LEA_y',])
#This code is used to remove the duplicate columns#

In [170]:
df=df.rename(columns={'LA_x':'LA'})

In [171]:
df

URN   LA  ... NOT_SUSTAINEDPER_DIS UNKNOWNPER_DIS
0     100001  201  ...                  NaN            NaN
1     100002  201  ...                  NaN            NaN
2     100003  201  ...                  NaN            NaN
3     100049  202  ...                 0.11           0.01
4     100050  202  ...                 0.04           0.01
...      ...  ...  ...                  ...            ...
6391  148135  391  ...                  NaN            NaN
6392  148179  845  ...                  NaN            NaN
6393  148185  925  ...                  NaN            NaN
6394  148186  808  ...                  NaN            NaN
6395  148187  330  ...                  NaN            NaN

[6396 rows x 35 columns]

# **Merge Ratings**

In [172]:
rat=pd.read_excel('/content/drive/MyDrive/OFSTED Data for Assignment/OFSTED Ratings.xlsx')

In [173]:
rat

Rating     URN
0     Unknown  100001
1     Unknown  100002
2     Unknown  100003
3        Good  100049
4        Good  100050
...       ...     ...
6391  Unknown  148135
6392  Unknown  148179
6393  Unknown  148185
6394  Unknown  148186
6395  Unknown  148187

[6396 rows x 2 columns]

In [174]:
data=pd.DataFrame.merge(df,rat,on='URN',how='left')

In [175]:
data

URN   LA GENDER  ... NOT_SUSTAINEDPER_DIS  UNKNOWNPER_DIS   Rating
0     100001  201  Girls  ...                  NaN             NaN  Unknown
1     100002  201  Mixed  ...                  NaN             NaN  Unknown
2     100003  201   Boys  ...                  NaN             NaN  Unknown
3     100049  202  Mixed  ...                 0.11            0.01     Good
4     100050  202  Girls  ...                 0.04            0.01     Good
...      ...  ...    ...  ...                  ...             ...      ...
6391  148135  391  Mixed  ...                  NaN             NaN  Unknown
6392  148179  845  Mixed  ...                  NaN             NaN  Unknown
6393  148185  925  Mixed  ...                  NaN             NaN  Unknown
6394  148186  808  Mixed  ...                  NaN             NaN  Unknown
6395  148187  330  Mixed  ...                  NaN             NaN  Unknown

[6396 rows x 36 columns]

# **Missing Values**

In [176]:
data.isnull().sum()

URN                        0
LA                         0
GENDER                     0
RELCHAR                  143
PNORG                    756
PNORB                    756
PSENELSE                 760
PSENELK                  758
PNUMEAL                 2172
PNUMENGFL               2172
PNUMUNCFL               2172
PNUMFSMEVER             2174
PNUMFSM                 2172
PERCTOT                 2158
PPERSABS10              2158
TPUP                    2448
ATT8SCR                 2448
P8PUP                   2448
P8MEA                   2448
P8CILOW                 2448
P8CIUPP                 2448
PTL2BASICS_94           2448
PTL2BASICS_95           2448
EBACCAPS                2448
PTEBACC_E_PTQ_EE        2448
PTEBACC_94              2448
PTEBACC_95              2448
PT5EM_94                2448
P8_BANDING              2863
OVERALL_DESTPER         2668
NOT_SUSTAINEDPER        2668
UNKNOWNPER              2668
OVERALL_DESTPER_DIS     2668
NOT_SUSTAINEDPER_DIS    2668
UNKNOWNPER_DIS

It can be seen above that the attributes 'URN', 'LA', 'GENDER' and 'Rating' do not contain any missing data. With regards to the other attributes' missing data, the UK government website reveals a number of things. Firstly, there are a number of schools that have opened relatively recently so information such as 'Rating', 'PNUMFSM' will not yet be available. Furthermore, there are cases where the school became an academy or changed type during the 2018 to 2019 academic year. In these cases, the 2019 data is found on the former school's page and, therefore, no data exists for 2020. Other instances exist whereby data related to Progress 8 such as 'P8PUP' and 'P8MEA' are not published if the schools are independent schools and independent special schools, or breakdowns by disadvantaged and other pupils for independent schools, independent special schools and non-maintained special schools. There are also schools on the list that do not have any data available for attributes such as 'ATT8SCR' and 'EBACCAPS' because the school or college did not enter any pupils or students for the relevant qualifications.

# **Handling Missing Data**

Deciding on what to do with the missing data depends on what kind of analysis will be carried out on the dataframe and what kind of answers we are looking for. 

# Option 1
The first option is to remove all rows that contain missing data. This would produce a dataframe that contains no missing data but has significantly less rows (schools).

In [177]:
data.isnull().any(axis=1).sum()

3187

As shown above, there are 3187 rows (schools) that have one or more values missing. This leaves 3209 schools that do not have any data missing. This is just over 50% of the schools which makes omitting rows with empty values a viable option.

# Option 2
Another option is to omit the columns that contain the most missing values such as 'P8_BANDING' and from there, remove the rows that contain missing data. This will produce a dataframe with more schools than the first option but less attributes to compare. For example:

In [178]:
df2=data.drop(columns=['PNUMEAL',
'PNUMENGFL',             
'PNUMUNCFL',               
'PNUMFSMEVER',           
'PNUMFSM',                 
'PERCTOT',            
'PPERSABS10',           
'TPUP',            
'ATT8SCR',             
'P8PUP',                
'P8MEA',                
'P8CILOW',              
'P8CIUPP',              
'PTL2BASICS_94',          
'PTL2BASICS_95',          
'EBACCAPS',            
'PTEBACC_E_PTQ_EE',       
'PTEBACC_94',          
'PTEBACC_95',          
'PT5EM_94',           
'P8_BANDING',
'OVERALL_DESTPER',
'NOT_SUSTAINEDPER',
'UNKNOWNPER',
'OVERALL_DESTPER_DIS',
'NOT_SUSTAINEDPER_DIS',
'UNKNOWNPER_DIS'])

In [179]:
df2

URN   LA GENDER            RELCHAR  ...  PNORB  PSENELSE  PSENELK   Rating
0     100001  201  Girls               None  ...    0.0       0.0      7.7  Unknown
1     100002  201  Mixed  Church of England  ...   58.7       0.0      5.5  Unknown
2     100003  201   Boys               None  ...  100.0       0.0     13.0  Unknown
3     100049  202  Mixed     Does not apply  ...   58.6       3.2     11.8     Good
4     100050  202  Girls     Does not apply  ...    4.0       1.3     10.2     Good
...      ...  ...    ...                ...  ...    ...       ...      ...      ...
6391  148135  391  Mixed     Does not apply  ...    NaN       NaN      NaN  Unknown
6392  148179  845  Mixed     Does not apply  ...    NaN       NaN      NaN  Unknown
6393  148185  925  Mixed               None  ...    NaN       NaN      NaN  Unknown
6394  148186  808  Mixed               None  ...    NaN       NaN      NaN  Unknown
6395  148187  330  Mixed               None  ...    NaN       NaN      NaN  Unknown

[6396 rows x 9 columns]

In [180]:
df2.isnull().any(axis=1).sum()

774

In [181]:
df2.dropna()

URN   LA GENDER            RELCHAR  ...  PNORB  PSENELSE  PSENELK   Rating
0     100001  201  Girls               None  ...    0.0       0.0      7.7  Unknown
1     100002  201  Mixed  Church of England  ...   58.7       0.0      5.5  Unknown
2     100003  201   Boys               None  ...  100.0       0.0     13.0  Unknown
3     100049  202  Mixed     Does not apply  ...   58.6       3.2     11.8     Good
4     100050  202  Girls     Does not apply  ...    4.0       1.3     10.2     Good
...      ...  ...    ...                ...  ...    ...       ...      ...      ...
6111  146640  394  Mixed               None  ...   93.8      79.2     20.8  Unknown
6113  146648  887  Mixed     Does not apply  ...   79.4       0.0     22.9  Unknown
6118  146697  937  Mixed               None  ...   53.5       0.7     13.0  Unknown
6119  146698  933  Mixed     Does not apply  ...   57.1     100.0      0.0  Unknown
6143  146825  845  Mixed               None  ...   50.9       0.9      5.7  Unknown

[5622 rows x 9 columns]

As shown above, by dropping columns that contain significant amounts of missing data, the number of rows (schools) containing missing data decreased from 3187 to 774. This means that there are 5622 schools to compare instead of 3209 (as in Option 1). However, only 9 columns remain meaning there are less attributes to compare.

# Option 3

Finally, if it is necessary for all schools to be compared, every column containg missing data can be removed such that the dataframe is reduced to the attributes 'URN', 'LA', 'GENDER' and 'Rating'.

In [182]:
df3=data.dropna(axis=1)

In [183]:
df3

URN   LA GENDER   Rating
0     100001  201  Girls  Unknown
1     100002  201  Mixed  Unknown
2     100003  201   Boys  Unknown
3     100049  202  Mixed     Good
4     100050  202  Girls     Good
...      ...  ...    ...      ...
6391  148135  391  Mixed  Unknown
6392  148179  845  Mixed  Unknown
6393  148185  925  Mixed  Unknown
6394  148186  808  Mixed  Unknown
6395  148187  330  Mixed  Unknown

[6396 rows x 4 columns]

This means that no rows (schools) need to be removed so all 6396 schools can be compared. However, only the 4 aforementioned attributes can be compared across the schools. 

In this instance, it would be prudent to go with option 1. Maintaining all the attributes will be useful for extensive analysis of each school and allows many different types of comparisons to be made. Although the dataframe will be made smaller, more than 50% of the schools will still remain and virtually every local authority is represented which can help give an idea of what might be expected from the schools that have missing data.

## Dataframe after Option 1

In [184]:
data1=data.dropna(axis=0)

In [185]:
data1

URN   LA GENDER  ... NOT_SUSTAINEDPER_DIS  UNKNOWNPER_DIS   Rating
3     100049  202  Mixed  ...                 0.11            0.01     Good
4     100050  202  Girls  ...                 0.04            0.01     Good
5     100051  202  Mixed  ...                 0.09            0.06     Good
6     100052  202  Mixed  ...                 0.13               0     Good
7     100053  202  Mixed  ...                 0.07            0.03     Good
...      ...  ...    ...  ...                  ...             ...      ...
6019  146232  856  Mixed  ...                  0.1               0  Unknown
6020  146235  860  Mixed  ...                    0               0  Unknown
6021  146237  926  Mixed  ...                 0.06            0.06  Unknown
6025  146250  909  Mixed  ...                 SUPP            SUPP  Unknown
6026  146255  936  Mixed  ...                  0.1               0  Unknown

[3209 rows x 36 columns]

In [186]:
data1.isnull().sum()

URN                     0
LA                      0
GENDER                  0
RELCHAR                 0
PNORG                   0
PNORB                   0
PSENELSE                0
PSENELK                 0
PNUMEAL                 0
PNUMENGFL               0
PNUMUNCFL               0
PNUMFSMEVER             0
PNUMFSM                 0
PERCTOT                 0
PPERSABS10              0
TPUP                    0
ATT8SCR                 0
P8PUP                   0
P8MEA                   0
P8CILOW                 0
P8CIUPP                 0
PTL2BASICS_94           0
PTL2BASICS_95           0
EBACCAPS                0
PTEBACC_E_PTQ_EE        0
PTEBACC_94              0
PTEBACC_95              0
PT5EM_94                0
P8_BANDING              0
OVERALL_DESTPER         0
NOT_SUSTAINEDPER        0
UNKNOWNPER              0
OVERALL_DESTPER_DIS     0
NOT_SUSTAINEDPER_DIS    0
UNKNOWNPER_DIS          0
Rating                  0
dtype: int64

The new dataframe now consists of 3209 rows (schools) and 36 columns (attributes) with no missing data. 

## **Handling Invalid Data**

In [219]:
(data1['URN'] < 0).all()

False

# **Sanity Checks**

In [188]:
((data1['PNUMEAL'] + data1['PNUMENGFL'] + data1['PNUMUNCFL']) >99.89).all()

True

PNUMEAL, PNUMENGFL, PNUMUNCFL values should all equal 100 across each row. The code above shows that they all equal at least 99.89. The reason for the use of 99.89 is because of rounding and decimal places that may cause some totals to not equal exactly 100. When 99.9 is used, some totals fall below this threshold. 

In [198]:
data1['UNKNOWNPER'] = data1['UNKNOWNPER'].astype(str)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [199]:
data1['NOT_SUSTAINEDPER'] = data1['NOT_SUSTAINEDPER'].astype(str)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [200]:
data1['OVERALL_DESTPER'] = data1['OVERALL_DESTPER'].astype(str)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [203]:
data1['UNKNOWNPER'] + data1['NOT_SUSTAINEDPER'] + data1['OVERALL_DESTPER']

3        0.010.090.9
4       0.020.040.94
5       0.050.080.86
6       0.010.080.91
7       0.010.040.95
            ...     
6019    0.010.040.95
6020       00.030.97
6021    0.010.020.97
6025             001
6026       00.090.91
Length: 3209, dtype: object

In [ ]:
data1['NOT_SUSTAINEDPER']

In [ ]:
data1['OVERALL_DESTPER']

Similarly, UNKNOWNPER, NOT_SUSTAINEDPER, OVERALL_DESTPER values should all equal 100 across each row. The code above shows that they all equal at least 99. The reason for the use of 99 is because of rounding and decimal places that may cause some totals to not equal exactly 100. 